### ISSUE TITLE
Store theme in project and restore it when launching it

### ISSUE THREAD

**Issue by BlackMagicCoding** [NationalSecurityAgency/ghidra#5336]
Hi all,

for my current project I have 3 separate Ghidra projects and code browsers open at the same time.
This is due to me having 3 different versions of an application and I have to merge information from 2 of those versions into the 3rd one.
To make it easier for my brain and eyes to distinguish which version I am currently looking at, I have created separate themes for each of those projects, which boils down to just different background colors.
The new theming support which was introduced with Ghidra 10.3 can totally do this, and I absolutely love it - it's a blessing for this tool!

At the moment the only thing making this bitter sweet is, that the theme seems to be set globally, meaning when I change it and launch any of those projects new, it will load with the theme last set.
My idea to improve this is now to save the theme information in the project and then load this theme when Ghidra launches it.
Having different projects and code browsers open with different themes at the same time is possible, but it requires me clicking a lot every time I launch these projects, which is quite a bit cumbersome (see following screenshot).

Any thoughts on this?

Kind regards,
an excited Ghidra newbie

**Comment by BlackMagicCoding** [comment]
[https://i.postimg.cc/50rHpr4k/2023-05-15-20-33-21.png](screenshot of 3 differently themed ghidra projects open at the same time)

**Comment by BlackMagicCoding** [comment]
![image](https://github.com/NationalSecurityAgency/ghidra/assets/57157966/04e50231-0eb0-4e70-9a69-1a048b2b5959)

**Comment by ghidragon** [comment]
I'll put in an internal ticket to give users the option of specifying the theme to be either a global property or a project specific property.

**Comment by BlackMagicCoding** [comment]
Heyo, just knocking on the door again.
Could you tell me if this is still under consideration?
I really love working with Ghidra, but yesterday I was working in the wrong project window again, because the themes have to be manually set for each project opened (and I was too lazy to set it once again 😅)

**Comment by ghidragon** [comment]
It is in our backlog of things to do/consider.  I looked at it briefly a while ago and all I remember is that it wasn't trivial to do. I don't remember now exactly what was the difficulty was, but I think it was more a conceptual issue. For example, what theme do you used if no project is open. I suppose it could be done as some sort of override. In other words , there would be a global theme that is stored the way it is now, but then maybe you can set a theme override for a project.

**Comment by ghidragon** [comment]
Actually, after reading you posts again, we believe that having separate preferences/settings for each of your running Ghidra instances would fix your use case.  To use this solution:

- Create a ghidraRun.bat file for each instance
- Add a line to the new file to specify an alternate preferences/settings dir (this is what we propose being able to do)
- Then any changes such as setting the theme will be saved in that specified settings dir

Each time you run Ghidra using that batch file, those particular preferences / theme will be used.

We think this is a more general solution to instances interfering with each other and should be fairly easy to implement.
Do you think this would satisfy your use case?

**Comment by BlackMagicCoding** [comment]
While not being perfect (I would have to keep all settings except Theme in sync) I would give it a go.
Could you lend me a hand?
I tried to understand it by looking at the .bat files (working on Windows), but I didn't really get where exactly I have to provide the alternate settings directory.
So far I have seen (and also read about) that there is a .ghidra folder in my home directory.
This is true and there is ghidra_11.0_PUBLIC inside.

**Comment by dragonmacher** [comment]
> I tried to understand it by looking at the .bat files (working on Windows), but I didn't really get where exactly I have to provide the alternate settings directory.

Just to be clear, Ghidra does not support this currently.    @ghidragon was suggesting a new feature we could for your use case.

> While not being perfect 

What sort of solution or workflow do you think would be the best for your situation?

**Comment by BlackMagicCoding** [comment]
> > I tried to understand it by looking at the .bat files (working on Windows), but I didn't really get where exactly I have to provide the alternate settings directory.
> 
> Just to be clear, Ghidra does not support this currently. @ghidragon was suggesting a new feature we could for your use case.
> 
> > While not being perfect
> 
> What sort of solution or workflow do you think would be the best for your situation?

Oops my bad ^^

The simplest solution I could think of is the following:
![image](https://github.com/NationalSecurityAgency/ghidra/assets/57157966/7d5fcdbb-0d41-4479-b849-a815641d3ef2)
Ghidra could be aware when going to the menu entry "Edit > Theme" - if a project is open, show an additional entry "Set project theme".
Show the same dialog as usual, but save the name of the chosen theme within the project.

This is the handling I would do when opening a project:
- check if a theme name was saved with the project
  - no
    - simply keep using the theme already set
  - yes
    - check if a theme with that name actually exists
      - no
        - tell the user that the theme could not be found
        - clear the saved theme name (optionally tell the user you did that)
        - keep using the theme already set
      - yes
        - load the specified theme, but without saving the potential theme change in the settings

I wouldn't make it too complex - just having a little help which avoids having to set/change the theme manually each and every time is already more than enough.
No need to make it extra complex in my opinion.
What do you think?

**Comment by dragonmacher** [comment]
This is close to what we were originally discussing internally.

What is not clear to us is how you are running multiple versions of Ghidra.  We made the assumption you were using multiple batch files, hence the suggestion.   How do you run your various Ghidra's?

**Comment by BlackMagicCoding** [comment]
I use the very same batch file on the same machine.
When I click the batch file 

> This is close to what we were originally discussing internally.
> 
> What is not clear to us is how you are running multiple versions of Ghidra. We made the assumption you were using multiple batch files, hence the suggestion. How do you run your various Ghidra's?

I click the same .bat file multiple times.
After the first time Ghidra complains that it can not open the project, because it is already open (guess it saves it as open project right away).
Once I dismissed the error message I simply open one of the other 2 projects.
Repeat once more to open the third one.

If there is a way to open a specific project by passing as command line parameter, I would definitely want to do that too, so I don't have to deal with those pesky messages 😅

**Comment by dragonmacher** [comment]
> If there is a way to open a specific project by passing as command line parameter, I would definitely want to do that too, so I don't have to deal with those pesky messages 😅

Yes.   You can edit the batch file and make the last parameter be the path to a project file.   For example, :

```
"${SCRIPT_DIR}"/support/launch.sh bg jdk Ghidra "${MAXMEM}" "" ghidra.GhidraRun "$@" /home/user/Ghidra/Projects/Project1.gpr
```

Then you can have multiple copies of your batch file, each pointing to its own project.

**Comment by BlackMagicCoding** [comment]
> > If there is a way to open a specific project by passing as command line parameter, I would definitely want to do that too, so I don't have to deal with those pesky messages 😅
> 
> Yes. You can edit the batch file and make the last parameter be the path to a project file. For example, :
> 
> ```
> "${SCRIPT_DIR}"/support/launch.sh bg jdk Ghidra "${MAXMEM}" "" ghidra.GhidraRun "$@" /home/user/Ghidra/Projects/Project1.gpr
> ```
> 
> Then you can have multiple copies of your batch file, each pointing to its own project.

Worked perfectly, thank you so much!
Maybe you could consider writing this as a comment in the ghidraRun.bat file as an example - might help a lot of other users as well.
I have seen in support/launch.bat that one can pass app-args, but what these app-args can be was always a mystery to me.

Now I wait patiently for the project theme support, provided that it is still considered an will be implemented.
Already solved a big annoyance with that .bat file fix 😄

**Comment by ryanmkurtz** [comment]
@BlackMagicCoding, you can currently add variations of the following line to your modified `ghidraRun.bat` files to have them use different user settings directories:

```
set VMARG_LIST=-Duser.home=C:\<dir>
```
Changing what Java thinks is your home directory is a bit of a hack, but it should result in Ghidra using unique locations for its user settings directories.

If you are running the latest `master` branch, or if you try this in Ghidra 11.1 when it gets released (no release date planned), you'll be able to use `-Dapplication.settingsdir=` instead.

**Comment by BlackMagicCoding** [comment]
Hi, just wanted to drop by, asking if this is still under consideration.
For now it is working by performing manual steps when starting up the projects every time.
It's certainly anything but high priority, but it would make it easier to have a theme stick to a project :)

I have tried both of these in my Ghidra 11.2.1, but the theme did not stick.
```
set VMARG_LIST=-Duser.home=C:\<dir>
set VMARG_LIST=application.settingsdir=C:\<dir>
```

Here is what I have in my .CMD startup file right now, maybe I wrote it the wrong way, or I am missing a part:
```
:: Ghidra launch

@echo off
setlocal

:: Maximum heap memory may be changed if default is inadequate. This will generally be up to 1/4 of 
:: the physical memory available to the OS. Uncomment MAXMEM setting if non-default value is needed.
::set MAXMEM=2G
set VMARG_LIST=-Dapplication.settingsdir=C:\Users\alphaomega\Documents\Ghidra\World-of-Warcraft-0.5.3.3368-enUS-WIN.rep
call "%~dp0support\launch.bat" bg jdk Ghidra "%MAXMEM%" "" ghidra.GhidraRun %* C:\Users\alphaomega\Documents\Ghidra\World-of-Warcraft-0.5.3.3368-enUS-WIN.gpr
```

**Comment by dragonmacher** [comment]
Closing this as it is something that we are not likely to get to in the foreseeable future.
